aten/src/ATen/native/UpSample.h

```diff
@@ -445,23 +458,46 @@ static inline void compute_source_index_and_lambda(
     lambda0 = static_cast<scalar_t>(1);
     lambda1 = static_cast<scalar_t>(0);
   } else {
-    using opmath_t = at::opmath_type<scalar_t>;
     const auto real_input_index =
         area_pixel_compute_source_index<opmath_t>(
             ratio, output_index, align_corners, /*cubic=*/false);
-    // when `real_input_index` becomes larger than the range the floating point
-    // type can accurately represent, the type casting to `int64_t` might exceed
-    // `input_size - 1`, causing overflow. So we guard it with `std::min` below.
-    input_index0 = std::min(static_cast<int64_t>(real_input_index), input_size - 1);
+    guard_index_and_lambda(real_input_index, input_size, input_index0, lambda1);
     int64_t offset = (input_index0 < input_size - 1) ? 1 : 0;
     input_index1 = input_index0 + offset;
-    lambda1 = std::min(
-      std::max(real_input_index - input_index0, static_cast<opmath_t>(0)),
-      static_cast<opmath_t>(1)
-    );
     lambda0 = static_cast<scalar_t>(1.) - lambda1;
   }
 }
 
+// It will not be used by data types other than BFloat16.
+template <typename scalar_in, typename scalar_out>
+void inline apply_grad_input(scalar_in* buffer_ptr, scalar_out* gin, int64_t size) {
+  constexpr bool is_bf16 = std::is_same<scalar_out, BFloat16>::value;
+  TORCH_CHECK(is_bf16,
+              "Upsample backward only support BFloat16 in the lower percision data types on CPU.")
+  constexpr bool is_fp32 = std::is_same<scalar_in, float>::value;
+  TORCH_CHECK(is_fp32,
+              "Upsample backward should use float as acc buffer for BFloat16 grad input on CPU.")
```

Comment:
What's the problem if you make the function `void inline apply_grad_input(float* buffer_ptr, BFloat16* gin, int64_t size)`?

Reply:
Although `scalar_in=float` and `scalar_out=BFloat16` will be only used at runtime, `scalar_in` and `scalar_out` may be float or double at compile time. It will raise "cannot convert ‘double*’ to ‘float*’ or "cannot convert ‘float*’ to ‘c10::BFloat16*’" at compile time when using `void inline apply_grad_input(float* buffer_ptr, BFloat16* gin, int64_t size)`.

Reply:
Does SFINAE work here so that you don't have to add runtime checks inside a general template function?

Reply:
Add a default template instead since SFINAE can't work because it still doesn't generate functions for data types other than BFloat16.